src/sentry/search/snuba/executors.py

```diff
@@ -71,18 +45,44 @@ def get_search_filter(search_filters, name, operator):
     return found_val
 
 
+@six.add_metaclass(ABCMeta)
 class AbstractQueryExecutor:
     """This class serves as a template for Query Executors.
     We subclass it in order to implement query methods (we use it to implement two classes: joined Postgres+Snuba queries, and Snuba only queries)
     It's used to keep the query logic out of the actual search backend,
     which can now just build query parameters and use the appropriate query executor to run the query
     """
 
+    TABLE_ALIAS = ""
+
+    @abstractproperty
+    def aggregation_defs(self):
+        """This method should return a dict of key:value
+        where key is a field name for your aggregation
+        and value is the aggregation function"""
+        raise NotImplementedError
+
+    @abstractproperty
+    def dependency_aggregations(self):
+        """This method should return a dict of key:value
+        where key is an aggregation_def field name
+        and value is a list of aggregation field names that the 'key' aggregation requires."""
+        raise NotImplementedError
+
+    @property
+    def empty_result(self):
+        return Paginator(Group.objects.none()).get_result()
+
+    @property
+    @abstractmethod
+    def dataset(self):
+        """"This function should return an enum from snuba.Dataset (like snuba.Dataset.Events)"""
+        raise NotImplementedError
+
+    @abstractmethod
     def query(
         self,
         projects,
-        retention_window_start,
-        group_queryset,
```

Comment:
you are still passing this parameters to the executor from SnubaSearchBackendBase.
Removing them makes tour subclasses not compatible with the abstract class. When you create an abstract class and let the subclasses implement an abstract method, the interface has to be compatible with the one of the abstract class (which may or may not be the exact same depending if your language is invariant-/covariant/contravariant - but in no case you can add or remove parameters.)

The goal of providing an abstract class and multiple implementations is taking advantage of polimorphism by making the callsite unaware of which specific implementation it would be using at a given moment. This implies that all implementation have to provide the same interface (which is defined in the abstract class). 
If the abstract class and the concrete class do not require the same parameters this concept is broken. (whether a concrete class would not use a parameter for that specific implementation is a smaller issue). I am surprised python even allows you to instantiate PostgresSnubaQueryExecutor. Definitively it would not pass the type checks of mypy.